src/npip.js

```diff
@@ -3,29 +3,73 @@
 "use strict";
 
 const fs = require("fs");
+const os = require("os");
 const path = require("path");
 const { findPackage, spawnPython } = require("./api");
-const { promisify } = require("./promisify");
 
-const stat = promisify(fs.stat);
-
-const installPip = (pkg) => {
-  return stat(path.join(pkg.dir, "python_modules")).catch(error => {
+const installPip = async (pkg) => {
+  try {
+    await fs.promises.stat(path.join(pkg.dir, "python_modules"));
+    return;
+  } catch (error) {
     if (error.code !== "ENOENT")
       throw error;
-    console.log("No python_modules directory; installing pip locally if needed.");
-    return spawnPython([path.join(__dirname, "get-pip.py"), "--user", "--quiet"], {
-      package: pkg,
-      interop: "status",
-      spawn: {
-        stdio: "inherit",
-      }
-    }).then(code => {
-      console.log("Successfully completed pip check.");
-      return code;
+  }
+
+  console.log("No python_modules directory; installing pip locally if needed.");
+  const opt = {
+    package: pkg,
+    interop: "status",
+    spawn: {
+      stdio: "inherit"
+    }
+  };
+
+  // Check if pip is already installed.
+  try {
+    await spawnPython(["-m", "pip", "-V"], opt);
+    return;
+  } catch { }
+
+  // Try to install via ensurepip.
+  try {
+    await spawnPython(["-m", "ensurepip", "--user"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch { }
+
+  // run get-pip.py.
+  try {
+    await spawnPython([path.join(__dirname, "get-pip.py"), "--user", "--quiet"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch { }
+
+  // Last resort: download pip and run it.
+  const tmpFolder = await fs.promises.mkdtemp(path.join(os.tmpdir(), 'npip-'));
+  const tmpFile = path.join(tmpFolder, "get-pip.py");
+
+  try {
+    await fetch("https://bootstrap.pypa.io/get-pip.py", {
+      method: "GET",
+      headers: {
+        "User-Agent": "nopy npip",
+      },
+    }).then(res => {
+      if (!res.ok)
+        throw new Error("Failed to download get-pip.py: " + res.statusText);
+      return res.body.pipe(fs.createWriteStream(tmpFile));
     });
-  });
-}
+    // run get-pip.py
+    await spawnPython([tmpFile, "--user", "--quiet"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch (error) {
+    console.error("Failed to install pip locally.\n" + JSON.stringify(error, null, 2));
+  } finally {
+    await fs.promises.unlink(tmpFolder);
```

Comment:
woupsie my bad

Reply:
> But that runs the risk of deleting files that were not intended

How so?

Reply:
Like if someone was hacking on the npip code and accidentally made tmpFolder point at /tmp, it would try to delete /tmp recursively. If you've run it and your /tmp directory survived then recursive version is fine :)

Reply:
It should be fine

```js
  const tmpFolder = await fs.promises.mkdtemp(path.join(os.tmpdir(), 'npip-'));
```
this will always be a random temp folder or at the very least an error